### PR TITLE
Torch binary features

### PR DESCRIPTION
**Pull request by skanjila**
# Code Pull Requests

Please provide the following:

- simple comment

### CODE REVIEW

File: ludwig/modules/loss_modules.py [hunk 1]
```diff
@@ -59,42 +59,42 @@ def forward(self, input, target):
...
+        max_entropy = torch.IntTensor(np.log(num_classes), dtype=torch.float32)
+        # clipping needed for avoiding log(0) = -inf
+        entropy_per_class = torch.maximum(-probabilities * torch.log(torch.clamp(probabilities, 1e-10, 1)), 0, )
+        entropy = torch.sum(entropy_per_class, -1)
+        penalty = (max_entropy - entropy) / max_entropy
+        return torch.mean(penalty)
+
+    def forward(self, input: torch.Tensor, target: torch.Tensor):
+        logits = target[LOGITS]
+        target = target.long()
```
Comment (w4nderlust) [inline ludwig/modules/loss_modules.py:77]: this is likely wrong. The type and the dict indexing fon't match, and targets and predictions seems flipped.
I would rename the erguments to be predictions and targets, and predictions has logits, not targets.

File: ludwig/modules/loss_modules.py [hunk 2]
```diff
@@ -59,42 +59,42 @@ def forward(self, input, target):
...
+        print(f'preds: {input.shape} {input.dtype} {input}')
+        print(f'target: {target.shape} {target.dtype} {target}')
+        output = self.loss_fn(input, target)
+        output = output.backward()
```
Comment (w4nderlust) [inline ludwig/modules/loss_modules.py:81]: this hsouldn't be here. Here we are just computing the loss, not computing the gradients

File: ludwig/modules/loss_modules.py (hunk 1 above)
Comment (skanjila) [reply ludwig/modules/loss_modules.py]: Fixed

File: ludwig/modules/loss_modules.py (hunk 2 above)
Comment (skanjila) [reply ludwig/modules/loss_modules.py]: Removed

File: ludwig/modules/loss_modules.py [hunk 3]
```diff
@@ -59,42 +59,41 @@ def forward(self, input, target):
...
+        entropy_per_class = torch.maximum(-probabilities * torch.log(torch.clamp(probabilities, 1e-10, 1)), 0, )
+        entropy = torch.sum(entropy_per_class, -1)
+        penalty = (max_entropy - entropy) / max_entropy
+        return torch.mean(penalty)
+
+    def forward(self, predictions: torch.Tensor, target: torch.Tensor):
+        logits = predictions[LOGITS]
+        target = target.long()
+        print(f'preds: {predictions.shape} {predictions.dtype} {input}')
+        print(f'target: {target.shape} {target.dtype} {target}')
```
Comment (w4nderlust) [inline ludwig/modules/loss_modules.py:79]: we can remove

File: ludwig/modules/loss_modules.py [hunk 4]
```diff
@@ -59,42 +59,41 @@ def forward(self, input, target):
...
-#             train_mean_loss += self.confidence_penalty * mean_penalty
-#
-#         return train_mean_loss
+class BWCEWLoss:
+    def __init__(self, **kwargs):
+        super().__init__()
+        self.loss_fn = nn.BCEWithLogitsLoss(**kwargs)
+
+    def mean_confidence_penalty(self, probabilities, num_classes):
+        max_entropy = torch.IntTensor(np.log(num_classes), dtype=torch.float32)
```
Comment (w4nderlust) [inline ludwig/modules/loss_modules.py:68]: Not really sure why we are creating an IntTensor but giving it type float32

File: ludwig/modules/loss_modules.py (hunk 3 above)
Comment (skanjila) [reply ludwig/modules/loss_modules.py]: removed

File: ludwig/modules/loss_modules.py (hunk 4 above)
Comment (skanjila) [reply ludwig/modules/loss_modules.py]: updated

File: ludwig/encoders/binary_encoders.py [hunk 5]
```diff
@@ -48,9 +48,10 @@ def __init__(
         super().__init__()
         logger.debug(' {}'.format(self.name))
 
-    def call(self, inputs, training=None, mask=None):
+    def forward(self, inputs, training=None, mask=None):
```
Comment (ShreyaR) [inline ludwig/encoders/binary_encoders.py:51]: Could you remove the `training` and `mask` arguments? I believe we aren't using them anymore in torch.

File: ludwig/encoders/binary_encoders.py [hunk 6]
```diff
@@ -48,9 +48,10 @@ def __init__(
...
         """
             :param inputs: The inputs fed into the encoder.
                    Shape: [batch x 1], type tf.float32
         """
-        return {'encoder_output': tf.cast(inputs, dtype=tf.float32)}
+        #inputs = inputs.type(torch.float32)
+        return {'encoder_output': inputs}
```
Comment (ShreyaR) [inline ludwig/encoders/binary_encoders.py:57]: Could you also add `input_shape` and `output_shape` properties for the class? Example here for [category encoders](https://github.com/ludwig-ai/ludwig/blob/ee17666addc269cb022272c55edba58b6dbaff97/ludwig/encoders/category_encoders.py#L85).

File: ludwig/features/binary_feature.py [hunk 7]
```diff
@@ -15,18 +15,18 @@
 # limitations under the License.
 # ==============================================================================
 import logging
-
 import numpy as np
 # import tensorflow as tf
```
Comment (ShreyaR) [inline ludwig/features/binary_feature.py:19]: Could you remove the commented out code from this file? Also at lines 20, 30, 136, etc.

File: ludwig/features/binary_feature.py [hunk 8]
```diff
@@ -231,7 +237,7 @@ def get_prediction_set(self):
 
     @classmethod
     def get_output_dtype(cls):
-        return tf.bool
+        return torch.bool
```
Comment (ShreyaR) [inline ludwig/features/binary_feature.py:240]: Could you also add an `input_shape` property for this feature?

File: tests/integration_tests/test_binary_feature.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
+import pytest
```
Comment (ShreyaR) [inline tests/integration_tests/test_binary_feature.py:1]: Thanks for adding this test!

Could you move this to `tests/ludwig/features`?

**Review by ShreyaR — COMMENTED** [review]
Thanks for porting this feature @skanjila! I left some minor comments..

File: ludwig/encoders/binary_encoders.py (hunk 5 above)
Comment (skanjila) [reply ludwig/encoders/binary_encoders.py]: removed

File: ludwig/encoders/binary_encoders.py (hunk 6 above)
Comment (skanjila) [reply ludwig/encoders/binary_encoders.py:56]: added

File: ludwig/features/binary_feature.py (hunk 7 above)
Comment (skanjila) [reply ludwig/features/binary_feature.py:19]: removed all commented out code

File: ludwig/features/binary_feature.py (hunk 8 above)
Comment (skanjila) [reply ludwig/features/binary_feature.py:221]: added

File: tests/integration_tests/test_binary_feature.py (hunk 9 above)
Comment (skanjila) [reply tests/integration_tests/test_binary_feature.py]: moved, although Jim's tests that follow this pattern live in the integration tests directory

Comment (ShreyaR) [reply tests/integration_tests/test_binary_feature.py]: I see. I can add a fix for that during the testing phase.

**Review by ShreyaR — APPROVED** [review]
lgtm! Feel free to merge the PR once @w4nderlust approves it as well.

File: ludwig/modules/loss_modules.py [hunk 10]
```diff
@@ -61,42 +61,39 @@ def forward(self, input, target):
...
-#             train_mean_loss += self.confidence_penalty * mean_penalty
-#
-#         return train_mean_loss
+class BWCEWLoss:
+    def __init__(self, **kwargs):
+        super().__init__()
+        self.loss_fn = nn.BCEWithLogitsLoss(**kwargs)
+
+    def mean_confidence_penalty(self, probabilities, num_classes):
+        max_entropy = torch.IntTensor(np.log(num_classes), dtype=torch.int32)
```
Comment (w4nderlust) [inline ludwig/modules/loss_modules.py:70]: change np.log with torch.log
